### ISSUE TITLE
combine multi json files and convert to a csv file

### ISSUE THREAD

**Issue by amarrano** [danforthcenter/plantcv#466]
**Start a discussion about anything you would like**
Discussions can be used to ask questions, discuss new ideas, or anything else.
Hi,

I'm a user of plantcv and python in general. I'm following the seed_size_plantcv_analysis.py script that I cloned from https://github.com/danforthcenter/plantcv-seeds. Also, I'm applying the seed image provided in the data folder of plantcv.

After editing the script based on the new updated pcv functions, I was able to get a json file for each analyze_object (i.e., seed) as follows:

for i in range(0, len(roi_objects)):
    if roi_obj_hierarchy[0][i][3] == -1:  # Checks if shape is a parent contour
                                        
        # Object combine kept objects
        obj, mask2 = pcv.object_composition(img, [roi_objects[i]], np.array([[roi_obj_hierarchy[0][i]]]))
        
        if obj is not None:
            analysis_image = pcv.analyze_object(img, obj, mask2)
            
            # Add custom output
            pcv.outputs.add_observation(
                    variable='filename',
                    trait='roi',
                    method='roi',
                    scale='int',
                    datatype=int,
                    value=i,
                    label='#')
            
            # At this point we have observations for one plant
            # We can write these out to a unique results file
            # Here I will name the results file with the ROI ID combined with the original result filename
            filename = args.result + "_" + str(i) + ".json" 
            # Save the existing metadata to the new file
            with open(filename, "w") as r:
                r.write(metadata)
            pcv.print_results(filename=filename)
            pcv.outputs.clear()

I would like to combine all the json files in a unique one, which then can be converted in a csv file using json2csv.

I'm trying to apply parallel.process_results but I get a preresult.json file with just {"entities": [], "variables": {}}, which means (as I understood by previous posted issues) no results. All json files are stored in a subdirectory.

Is process_results the right function to use in this case? 

Thank you.

**Comment by HaleySchuhl** [comment]
Hi @amarrano ,

You are correct that using `plantcv.parallel.process_results` will combine and format data into a json file that can get used for input in `json2csv`. 

We have discussed the changes we'd like to make to streamline this process of retrieving data in a format appropriate for downstream analysis. #422 The `json2csv` function was originally designed to run on the combined json file that is output after running PlantCV in parallel over a dataset of images. In the future we want to make the intermediate step of processing results output after just one image will be compatible with the input for the `json2csv` function. 

I'm not sure where the disconnect is with the example you've described. Did you confirm that there is data stored in the results files created by `pcv.print_results`? Our collaborator from the issue tagged here ran into issues since there was different syntax on Windows; what operating system and version of PlantCV are you using while getting this error? Could you please provide these details, and what you ran from the command line that resulted in no results getting transferred into the csv file?

**Comment by amarrano** [comment]
Thank you for your fast replay. I'm attaching four results files (.txt just because the json is not supported here as an extension) generated from my command. I get a different json file for each seed in the single image (totally 69 seeds).

I'm using Windows and plantcv 3.6.0







[seed-size-Oct16_2019_results_2.txt](https://github.com/danforthcenter/plantcv/files/3741107/seed-size-Oct16_2019_results_2.txt)
[seed-size-Oct16_2019_results_3.txt](https://github.com/danforthcenter/plantcv/files/3741108/seed-size-Oct16_2019_results_3.txt)
[seed-size-Oct16_2019_results_4.txt](https://github.com/danforthcenter/plantcv/files/3741109/seed-size-Oct16_2019_results_4.txt)
[seed-size-Oct16_2019_results_0.txt](https://github.com/danforthcenter/plantcv/files/3741110/seed-size-Oct16_2019_results_0.txt)
[seed-size-Oct16_2019_results_1.txt](https://github.com/danforthcenter/plantcv/files/3741111/seed-size-Oct16_2019_results_1.txt)

**Comment by nfahlgren** [comment]
Thanks @amarrano, I can confirm that this is a bug. When we run `plantcv-workflow.py` to process data and run a user workflow, the intermediate JSON files have metadata in addition to the PlantCV observations data, but when the intermediate files are created by a user workflow directly outside of `plantcv-workflow.py` there is not necessarily metadata. The current version of process_results is looking for the keys to the metadata dictionary (`for vars in obs["metadata"].keys():`) but since the dictionary isn't there the keys() method returns an error.

There are two relatively easy fixes we can apply. I think one option is to update `process_results` so it doesn't rely on the keys method:

```
for vars in obs["metadata"]:
```

This will loop over the keys just like with the use of the keys method, but only if the dictionary is defined.

Alternatively, or additionally, we could update `print_results` to add an empty metadata dictionary if it doesn't exist.

**Comment by amarrano** [comment]
Thank @nfahlgren. To see if I understood, my json files created with print_results can't be processed by process_results because there is no metadata. 
 
So, in my case, will I need to develop the workflow and then apply process_results (even if in this case it refers to only one image)? Also, which are the exact argument of process_results? I'm new and I can't find any info in the tutorials.

All my json files are saved in ..\Linux-amarrano\myNotebook\seed-size-Oct16_2019\output_json, and when I run parallel.process_results('','presults.json') in my notebook (see command below), it saves the presults.json in myNotebook folder. 

When I run  parallel.process_results('seed-size-Oct16_2019/output_json/','presults.json'), I got the following error, which I think is what you are saying:

KeyError                                  Traceback (most recent call last)
<ipython-input-398-5f14d98c9dc2> in <module>()
----> 1 parallel.process_results('seed-size-Oct16_2019/output_json/','presults.json')

/home/amarrano/.local/lib/python2.7/site-packages/plantcv/parallel/process_results.pyc in process_results(job_dir, json_file)
     41                     data["entities"].append(obs)
     42                     # Keep track of all metadata variables stored
---> 43                     for vars in obs["metadata"].keys():
     44                         data["variables"][vars] = {"category": "metadata", "datatype": "<class 'str'>"}
     45                     # Keep track of all observations variables stored

KeyError: 'metadata'

Did I used the function correctly?

**Comment by nfahlgren** [comment]
Hi @amarrano, that's correct but I think we have the fix staged in pull request #467. Once that is merged `process_results` should work on your data without running the parallel processing program.

You bring up another good point which is that I forgot to add documentation for the functions in the `parallel` subpackage. Most of the functions in the subpackage are not useful to run on their own (they are all used by `plantcv-workflow.py`, but process_results is one that is, and it should all be documented anyhow.

The inputs to process_results are `job_dir`, which is the path to the directory that contains the intermediate JSON files (1 per seed in your case), and `json_file`, which is the output combined JSON file. When you ran `parallel.process_results('seed-size-Oct16_2019/output_json/','presults.json')` you did use it correctly, and the KeyError you get is the bug that needs to be fixed.

**Comment by amarrano** [comment]
Got it. Thank you for your explanations. I will then wait for the fixed version of procss_results.

In the meantime, I can try to work on converting the notebook in workflow and then apply the parallel subpackage to convert the json files for json2csv.

**Comment by nfahlgren** [comment]
Sorry, my pull request automatically closed this issue.

**Comment by nfahlgren** [comment]
How did you install PlantCV? Just asking because we merged the bug fix but it's not pushed out to PyPI (pip) or Bioconda (conda) yet.

**Comment by amarrano** [comment]
I think I did it usia PyPI, but I also cloned it from github.

**Comment by nfahlgren** [comment]
If you update your local GitHub clone and reinstall from it then the bug fix should will be applied and it should work now, though let us know if not. We can also push a new bugfix release to PyPI and Bioconda tomorrow if needed.

**Comment by amarrano** [comment]
Ok, thank you. I will give it a try and let you know.

**Comment by amarrano** [comment]
Hi,

I have tried to update my local GitHub as described here https://plantcv.readthedocs.io/en/latest/updating/, however, when I run again process_results I was getting the same error as yesterday.

I also tried to uninstall and install again the plantcv package, but I get the following error:

from pip._internal.main import main
ImportError: No module named _internal.main.

I tried to upgrade pip but still the same error. I remember that I got the same problem the first time and then I installed pcv using pip.
Today the following command worked, which I don't know if it is the right one:
python -m pip install --user plantcv

In any case, the function process_results doesn't work for me.

Any help how to fix the installation using git?

**Comment by HaleySchuhl** [comment]
Hi @amarrano ,

Since we haven't done a new release with the update to `process_results` there won't be an update available on pip yet. Are you running `process_results` from the terminal or from a Jupyter notebook? If from Jupyter could you please run `pcv.__version__`, or if from the terminal please run `pip freeze` and confirm the version of PlantCV you are running.

**Comment by amarrano** [comment]
PlantCV '3.6.1'. I don't see plantcv in the list generated with pip freeze.

I think I did something wrong with the installation, or there is something limiting me to install correctly the package by cloning the github folder.

**Comment by nfahlgren** [comment]
Hi @amarrano, 3.6.1 indicates it is still the release version and not the version with the updated `process_results`, which should be `3.6.1+2.g60129a5`.

If I understand how it's setup on your system correctly, then I think the following should work:

On the command line:
```bash
# Change to the plantcv GitHub directory (your path is likely different)
cd github/plantcv
# Make sure the GitHub clone is up to date
git pull
# Install the new version of PlantCV
python setup.py install --user
```

**Comment by amarrano** [comment]
Ok, I was able to run these commands, but then when I open my notebook and import plantcv, it still tells me that I have 3.6.1

**Comment by nfahlgren** [comment]
In your Jupyter notebook can you send me the output of this command?

`print(pcv.__file__)`

**Comment by amarrano** [comment]
/home/amarrano/.local/lib/python2.7/site-packages/plantcv/plantcv/__init__.pyc

**Comment by nfahlgren** [comment]
You might need to run `pip uninstall plantcv` and then rerun the `python setup.py install --user` command. I'm thinking there is both the PyPI version and the GitHub version coexisting and your notebook is grabbing the wrong one.

**Comment by amarrano** [comment]
With pip uninstall plantcv I get the following:
WARNING: Skipping plantcv as it is not installed.

I will try to uninstall and delete the cloned folder, and then retry to install by github.

**Comment by amarrano** [comment]
Ok I was able to remove plantcv 3.6.1 using python -m pip uninstall plantcv

now in my notebook if I run pcv.__version__ I get:
'3.6.1+2.g60129a5'

but then when I do:
parallel.process_results('seed-size-Oct16_2019/output_json/','presults.json')

I get again:
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-22-5f14d98c9dc2> in <module>()
----> 1 parallel.process_results('seed-size-Oct16_2019/output_json/','presults.json')

/home/amarrano/.local/lib/python2.7/site-packages/plantcv-3.6.1+2.g60129a5-py2.7.egg/plantcv/parallel/process_results.pyc in process_results(job_dir, json_file)
     41                     data["entities"].append(obs)
     42                     # Keep track of all metadata variables stored
---> 43                     for vars in obs["metadata"]:
     44                         data["variables"][vars] = {"category": "metadata", "datatype": "<class 'str'>"}
     45                     # Keep track of all observations variables stored

KeyError: 'metadata

**Comment by nfahlgren** [comment]
I was mistaken about the fix, I think we are going to need to update the `print_results` function to add in an empty `metadata` dictionary to fix this issue.

I do see one quick fix on your end that will not require a new version of PlantCV for now. If you comment out this code:

```
# Save the existing metadata to the new file
        with open(filename, "w") as r:
            r.write(metadata)
```

And delete all the current JSON files then `print_results` will create the JSON files with an empty `metadata` dictionary, which means that `process_results` will work.

**Comment by amarrano** [comment]
Ok, I will give it a try. Thank you.

**Comment by amarrano** [comment]
@nfahlgren Thank for your suggestions. They worked fine and I was able to merge the different json files in only one.

Now I have another problem converting the merged json file in csv. 
From my terminal:

~/plantcv/plantcv/utils$ plantcv-utils.py json2csv
plantcv-utils.py: command not found

**Comment by nfahlgren** [comment]
`plantcv-utils.py` is in the main plantcv GitHub directory, not the utils directory.

**Comment by amarrano** [comment]
Sorry, I haven't noticed it!

It runs fine but it generates empty files. I'm attaching the json file (as a zip because supported format) I want to convert in csv.
[presults.zip](https://github.com/danforthcenter/plantcv/files/3751773/presults.zip)

**Comment by nfahlgren** [comment]
Thanks @amarrano, could you possibly send me a zip file of the individual JSON files (prior to merging with process_results)? The combined JSON file is missing some information and I can do some testing.

**Comment by amarrano** [comment]
Yes, sure. See attached.
[output_json.zip](https://github.com/danforthcenter/plantcv/files/3751823/output_json.zip)

I used the following code to generate them:

# The result file should exist if plantcv-workflow.py was run
#if os.path.exists(args.result):
    # Open the result file
#    results = open(args.result, "r")
    # The result file would have image metadata in it from plantcv-workflow.py, read it into memory
#    metadata = results.read()
    # Close the file
#    results.close()
    # Delete the file, we will create new ones
#    os.remove(args.result)
#else:
    # If the file did not exist (for testing), initialize metadata as an empty string
#    metadata = "{}"

# Gets the area of each seed, saved in analysis_image
#table = []
for i in range(0, len(roi_objects)):
    if roi_obj_hierarchy[0][i][3] == -1:  # Checks if shape is a parent contour
                                        # roi_obj_hierarchy is a list of arrays - 
                                        #with this command they are selecting the third element of the 
                                        #i elements of the first (unique) array (0)

        # Object combine kept objects
        obj, mask2 = pcv.object_composition(img, [roi_objects[i]], np.array([[roi_obj_hierarchy[0][i]]]))
        
        if obj is not None:
            analysis_image = pcv.analyze_object(img, obj, mask2)
            
            # Add custom output
            #pcv.outputs.add_observation(
                   # variable='filename',
                   # trait='roi',
                   # method='roi',
                   # scale='int',
                   # datatype=int,
                   # value=i,
                   # label='#')
            
            # At this point we have observations for one plant
            # We can write these out to a unique results file
            # Here I will name the results file with the ROI ID combined with the original result filename
            filename = args.result + "_" + str(i) + ".json" 
            # Save the existing metadata to the new file
            #with open(filename, "w") as r:
            #    r.write(metadata)
            pcv.print_results(filename=filename)
            # The results are saved, now clear out the observations so the next loop adds new ones for the next plant
            pcv.outputs.clear()

**Comment by amarrano** [comment]
Also, I rearranged my script from the original one based on the new updates. 
The original script, however, was not creating the intermediate files, but it was appending the results to a table. 

Howver, I was not able to extract the metadata from this table following the script. 
# Gets the area of each seed, saved in shape_data
    shape_header = []
    table = []
    for i in range(0, len(roi_objects)):
        if roi_obj_hierarchy[0][i][3] == -1:  # Checks if shape is a parent contour

            # Object combine kept objects
            device, obj, mask2 = pcv.object_composition(img, [roi_objects[i]], np.array([[roi_obj_hierarchy[0][i]]]),
                                                        device, debug)
            if obj is not None:
                device, shape_header, shape_data, shape_img = pcv.analyze_object(img, rgb_img, obj, mask2, device,
                                                                                 debug)
                if shape_data is not None:
                    table.append(shape_data)

    # Finds the area of the size marker in pixels and saves to "marker data"
    device, marker_header, marker_data, analysis_images =\
        pcv.report_size_marker_area(img, 'rectangle', device, debug, "detect", 3525, 850, -200, -1700, "black",
                                    "light", "h", 120)
    # shape_header.append("marker_area")

    # Saves seed and marker shape data results to file
    metadata = open(posixpath.join(outdir, outfile), 'r').read()
    os.remove(posixpath.join(outdir, outfile))

    for seed, row in enumerate(table):
        prefix = posixpath.join(outdir, outfile[0:-4])
        results = open(prefix + '_' + str(seed + 1) + '.txt', 'w')
        results.write(metadata)
        results.write('\t'.join(map(str, shape_header)) + '\n')
        # row.append(marker_data[1])
        results.write('\t'.join(map(str, row)) + '\n')
        results.write('\t'.join(map(str, marker_header)) + '\n')
        results.write('\t'.join(map(str, marker_data)) + '\n')
        results.close()

Do you think changing this approach will be easier than my strategy? If yes, I can I extract the observation from the table? When I save it, i just got the list of arrays with their indexes.

**Comment by nfahlgren** [comment]
Hi @amarrano, the code in your last post has some of the old syntax (3 outputs for `analyze_object` for example). In recent PlantCV releases we no longer return a data table along with the `analyze_object` output, the results are captured in a background observations object that `print_results` will save to a file. You can extract measurements directly from the observations within the workflow however, if needed. An example for getting the area of an object:

```
pcv.outputs.observations['pixel_area']['value']
```

If you run `process_results` on the contents of the output_json directory do you get the combined JSON file from above that did not work with json2csv?

**Comment by amarrano** [comment]
Hi @nfahlgren,

Yes, I got the combined Json file from the directory I sent you.

I understood that, since the functions changed, that script was not going to work anymore. I was just wondering if there was an alternative way to export my results to save each roi separately and then  combine the json files.

**Comment by HaleySchuhl** [comment]
Hi @amarrano ,

One alternative available right now, which avoids the need to combine multiple json files per image, is to store the information for all ROI's using the `add_observation` method just like you are already doing here:

```
# Add custom output
pcv.outputs.add_observation(
                variable='filename',
                trait='roi',
                method='roi',
                scale='int',
                datatype=int,
                value=i,
                label='#')
```

This might only be worth it if you are interested in a subset of the shape characteristics measured by the `pcv.analyze_object` function since it is verbose. 

```
area_table = []        # Store the PlantCV measurements for each seed in a table
perimeter_table = []
for i in range(0, len(roi_objects)):
    if roi_obj_hierarchy[0][i][3] == -1:  # Checks if shape is a parent contour

        # Object combine kept objects
        obj, mask2 = pcv.object_composition(img, [roi_objects[i]], np.array([[roi_obj_hierarchy[0][i]]]))
    
        if obj is not None:
            analysis_image = pcv.analyze_object(img, obj, mask2)
    
        if analysis_image is not None:
            area_table.append(pcv.outputs.observations['area']['value'])
            perimeter_table.append(pcv.outputs.observations['perimeter']['value'])

pcv.outputs.clear()
num_seeds = len(area_table)
seed_area_labels = [i for i in range(0, num_seeds)]

# Add area data
pcv.outputs.add_observation(variable='seed area', trait='list of seed area in a single image ',
                            method='pcv.analyze_object', scale='pixels', datatype=list,
                            value=area_table, label=seed_area_labels)
# Add perimeter data
pcv.outputs.add_observation(variable='seed perimeter', trait='list of seed perimeter in a single image ',
                            method='pcv.analyze_object', scale='pixels', datatype=list,
                            value=perimeter_table, label=seed_area_labels)

pcv.outputs.add_observation(variable='number_of_seeds_in_image', trait='',
                            method='pcv.object_composition', scale='', datatype=int,
                            value=num_seeds, label='seeds')

pcv.print_results(filename=filename)

```

Whether you make 1 json file per image or multiple the `process_results` step is still needed under the current version of PlantCV to get the formatted json file that is the input for `json2csv`. I downloaded the presults.json file you attached in this image I ran the following in the terminal:

` (plantcv) Downloads hschuhl$ plantcv-utils.py json2csv -j presults.json -c result-table.txt`

I was able to recreate the issue of creating an empty csv file from the presults.json file you attached here. I think I figured out the problem but I don't yet have a solution. It seems like usually the json files created will include output like `"datatype": "<class 'str'>"` or `"datatype": "<class 'int'>",` etc. However, in your output we have the same format but many instances of `"datatype": "<type 'int'>"`. The reason the `json2csv` function is running with no error, but also printing little to nothing to the 2 csv files that get created, is because the function looks for very specific output in each json file. 

Lines 21-24 of pcv.utils.json2csv: 
```
            elif var["datatype"] in ["<class 'bool'>", "<class 'int'>", "<class 'float'>", "<class 'str'>"]:
                scalar_vars.append(key)
            elif var["datatype"] in ["<class 'list'>"]:
                multi_vars.append(key)
``` 
Again, I'm not sure why the output in your json files is different but it seems feasible to update the `json2csv` function to be flexibility in the way it sorts through data while making the csv file.

**Comment by amarrano** [comment]
Hi @HaleySchuhl,

Thank you for your help and suggestions. I will try both approaches. 
Can I get the new, edit json2csv function by updating plantcv with github?

**Comment by HaleySchuhl** [comment]
Not yet but we can try to get a patch sometime today! What version of Python are you using? 

For the time being, this is how I got `json2csv` to cooperate:

```
import os
import json

def json2csv(json_file, csv_file):
    if os.path.exists(json_file):
        # If the JSON file exists open it for reading
        json_data = open(json_file, "r")
        data = json.load(json_data)
        # If the data is JSON but it does not have the components we expect from PlantCV raise an error
        if "variables" not in data or "entities" not in data:
            raise ValueError("Invalid JSON file: {0}".format(json_file))

        # Split up variables
        meta_vars = []
        scalar_vars = []
        multi_vars = []
        for key, var in data["variables"].items():
            if var["category"] == "metadata":
                meta_vars.append(key)
                
            elif var["datatype"] in ["<class 'bool'>", "<class 'int'>", "<class 'float'>", "<class 'str'>",
                                    "<type 'bool'>", "<type 'int'>", "<type 'float'>", "<type 'str'>"]:
                scalar_vars.append(key)
            elif var["datatype"] in ["<class 'list'>", "<type 'list'>"]:
                multi_vars.append(key)

        # Create a CSV file of single-value traits
        csv = open(csv_file + "-single-value-traits.csv", "w")

        # Build the single-value variables output table
        csv.write(",".join(map(str, meta_vars + scalar_vars)) + "\n")
        for entity in data["entities"]:
            row = []
            # Add metadata variables
            for var in meta_vars:
                print(var)
                obs = entity[data["variables"][var]["category"]]
                if var in obs:
                    row.append(obs[var]["value"])
                else:
                    row.append("NA")
            # Add scalar variables
            for var in scalar_vars:
                obs = entity[data["variables"][var]["category"]]
                if var in obs:
                    row.append(obs[var]["value"])
                else:
                    row.append("NA")
            csv.write(",".join(map(str, row)) + "\n")
        # Close the CSV file
        csv.close()

        # Create a CSV file of multi-value variables
        csv = open(csv_file + "-multi-value-traits.csv", "w")
        csv.write(",".join(map(str, meta_vars + ["trait", "value", "label"])) + "\n")
        for entity in data["entities"]:
            meta_row = []
            # Add metadata variables
            for var in meta_vars:
                obs = entity[data["variables"][var]["category"]]
                if var in obs:
                    meta_row.append(obs[var]["value"])
                else:
                    meta_row.append("NA")
            # Add multi-value variables
            for var in multi_vars:
                obs = entity[data["variables"][var]["category"]]
                if var in obs:
                    if obs[var]["label"] != "none":
                        for i in range(0, len(obs[var]["value"])):
                            row = [var]
                            row.append(obs[var]["value"][i])
                            row.append(obs[var]["label"][i])
                            csv.write(",".join(map(str, meta_row + row)) + "\n")
                else:
                    csv.write(",".join(map(str, meta_row + [var, "NA", "NA"])) + "\n")
        csv.close()
    else:
        # If the file does not exist raise an error
        raise IOError("File does not exist: {0}".format(json_file))
```

**Comment by amarrano** [comment]
I'm using Python 2.7.15+.

Then, should I change the json2csv function in the converters.py script in ~/plantcv/plantcv/utils?

**Comment by amarrano** [comment]
I changed the converters.py script as you suggested and it worked! 

I'm going to try also your suggestion of using outputs.add_observations to see if the JSON file will be still successfully converted in a CSV with the new json2csv function.

Thank you.

**Comment by HaleySchuhl** [comment]
I'm glad to hear you're able to get your data into csv files. I have the patch added to a branch that will get added in before the next PlantCV release. I'd definitely recommend updating once the next PlantCV release is done. Feel free to reopen this issue with any other questions or suggestions related to `process_results` or `json2csv`!

**Comment by amarrano** [comment]
Hi,

Thank to your suggestions I was able to have my first workflow working fine. I converted it in a python script and I tried to parallelize it, using multiple copies of the same image (still the seed image provided with plantcv). 

However, the output json file created with plantcv-workflow.py doesn't have results (e.g. seed areas), as, of course, the two csv files generated with json2csv.

I don't know if my bash script has something wrong, or if the results of my workflow are not read well from plantcv-workflow.py.

I'm attaching the folder including the input images, the json file from plantcv-workflow.py, the two cvs files, and my python and bash scripts. 

As a solution, I'm thinking to add a .py script that uses parallel.process_results to merge the json files obtained from each image into a single one to the bash script, but I was wondering if there is a better solution.

Thank you,

Annarita
[Issue.zip](https://github.com/danforthcenter/plantcv/files/3769107/Issue.zip)

**Comment by HaleySchuhl** [comment]
Hi @amarrano ,

I plan to look at this more carefully but on first glance I think the issue might have to do with line 156 of your seedSize_multipleJSON-files.py python script. Users that are running PlantCV in parallel shouldn't need to explicitly run `parallel.process_results` at any point in their workflow since it is part of `plantcv-workflow.py`. I will get back after I've gotten the chance to do more testing.

**Comment by amarrano** [comment]
Hi @HaleySchuhl,

I will give it a try without the parallel.process_results in the seedSize_multipleJSON-files.py python script and let you know.

Thank you.

**Comment by amarrano** [comment]
The tables are still without results even if without using parallel-process_results in the seedSize_multipleJSON-files.py python script.

**Comment by amarrano** [comment]
Hi all,

I tried to parallelize also the verbose workflow where I save each metric (e.g., area, perimeter, etc) in different tables (as @HaleySchuhl suggested before), but still, the final json, when converted to the csv, generates no data (even silencing the parallel.process_results() inside the workflow).

I was able to solve this by calling, in a unique sh script: first, the workflow that generates multiple JSONs files (one json file per seed in each image); then, parallel.process_results() in a separated python script; and then, plantcv-utils.py. 

I don't know if it is the most efficient solution, but it works and I can use it on my data for now. 
If you are able to identify the problem in my scripts that doesn't create a final json file with the metadata, then I will be happy to make my script more efficient.

Thank you,

Annarita

**Comment by nfahlgren** [comment]
Hi @amarrano, sorry for the delay! 

I modified your workflow script to get things working. Admittedly what I did is a bit hacky at the moment, @HaleySchuhl and I discussed some modifications to PlantCV we can start working on to make this a lot easier in the near future. For now here's a summary of what I did:

* Import the `json` module
* Added two helper functions to read metadata from the input results file, if it exists (works if it does not), and a function to create a per seed new JSON file.
* I commented out the blocks where the working directory is changed and moving the JSON files around as that should not be needed in the updated version.
* The size marker analysis was moved up before the analysis loop.
* Before the analysis for loop I used one of the helper functions to read the results file and delete it.
* After `analyze_object` I added a "custom" observation to add the marker area to the results for each seed (for convenience in downstream analysis).
* Used the second helper function to create a new results file (JSON) for each seed and then used `print_results` to add observations to it.

I tested the script on the data you sent using the following command (note we just updated `plantcv-workflow.py` and it has a new `timestampformat` input):

```bash
plantcv-workflow.py \
-d ./images \
-a filename \
-p ./seedSize_multipleJSON_files.py \
-j seed_data.json \
-f plantbarcode,timestamp,measurementlabel,imgtype \
-t jpg \
-s '%Y-%m-%d'
```

[seedSize_multipleJSON_files.py.zip](https://github.com/danforthcenter/plantcv/files/3781705/seedSize_multipleJSON_files.py.zip)

**Comment by amarrano** [comment]
Hi @nfahlgren,

I tried your script and it worked! It is much better than my solution because it doesn't save all the intermediated json files, saving a lot of memory!

Thank you so much for your help.

Annarita

**Comment by nfahlgren** [comment]
Fantastic!

**Comment by amarrano** [comment]
Hi there,

I'm testing my workflow for seed size on a test image of my seed (rice).
I modified a bit the script and I was able to fill in just the rice seeds and not the awns, which will avoid to clean the seed before the phenotyping. 

However, when I colored the seeds (objects) differently (a step right before the for loop for analyzing each seed separately), I noticed that some pixels of one seed are colored differently (you can notice them by zooming in), and are considered as different objects. Therefore, they are analyzed separately from the analyze_objects function.
I guess this is related to the variability of colors in each seed. 

The first solution that I thought was to fill objects with an area included in a specific interval, not just smaller than a certain number. Is there a way to do it in PlantCV? 

I'm attaching the commands and the images
[rice_Test_EpsonV600.zip](https://github.com/danforthcenter/plantcv/files/3795797/rice_Test_EpsonV600.zip)
.

args = options()
pcv.params.debug = args.debug

#let's safe the args.directory so that we can go back anytime:
originalDir = os.getcwd()

#read in image
img, imgpath, imgname = pcv.readimage(filename=args.image)

mask, masked_img = pcv.threshold.custom_range(rgb_img=img1, lower_thresh=[0,0,155], upper_thresh=[255,255,255], channel='LAB')

fill_image = pcv.fill(mask, 200) 
id_objects, obj_hierarchy = pcv.find_objects(img, fill_image)

roi1, roi_hierarchy= pcv.roi.rectangle(img=img1, x=150, y=800, h=2250, w=2250)
roi_objects, roi_obj_hierarchy, kept_mask, obj_area = pcv.roi_objects(img=img, roi_contour=roi1, roi_hierarchy=roi_hierarchy, object_contour=id_objects, 
                                                                      obj_hierarchy=obj_hierarchy, roi_type='partial')

# Randomly colors the individual seeds
img_copy = np.copy(img)
for i in range(0, len(roi_objects)):
    rand_color = pcv.color_palette(1)
    cv2.drawContours(img_copy, roi_objects, i, rand_color[0], -1, lineType=8, hierarchy=roi_obj_hierarchy)
image_name = extractName(args.image) + "_seedColors.jpg"
pcv.print_image(img_copy, image_name)#save the picture
pcv.outputs.clear()

**Comment by HaleySchuhl** [comment]
HI @amarrano ,

I'm going to open a new issue with this!
